matching/leet.go

```diff
@@ -52,15 +54,17 @@ func (lm l33tMatch) Matches(password string) []*match.Match {
 }
 
 func translate(password string, sub map[string]string) string {
-	var res string
-	for _, s := range password {
-		if v, ok := sub[string(s)]; ok {
-			res = res + v
-		} else {
-			res = res + string(s)
+	return strings.Map(func(r rune) rune {
```

Comment:
Could you please add a unit test for translate in leet_test.go with the relevant cases ? I'd like to understand better why the translate function might return more bytes than the original (I think it should never be the case since we only should replace ascii runes)

Reply:
I think TestsCornerCases has two examples where translate returns more bytes than the original. Could do with a comment or be broken out into a named test for translate.

Reply:
A better explanation: the problems occur when you have a string which is not valid utf8. As you range over the string, the rune value returned may be 0xfffd which is encoded in 3 bytes.

See https://golang.org/ref/spec#For_statements:
>For a string value, the "range" clause iterates over the Unicode code points in the string starting at byte index 0. On successive iterations, the index value will be the index of the first byte of successive UTF-8-encoded code points in the string, and the second value, of type rune, will be the value of the corresponding code point. If the iteration encounters an invalid UTF-8 sequence, the second value will be 0xFFFD, the Unicode replacement character, and the next iteration will advance a single byte in the string.

Therefore, for each invalid byte value which isn't valid utf8, s will be 0xfffd and len(string(s)) is 3.